docs/DesignDocs/oe_log_message()_callback_proposal.md

```diff
@@ -0,0 +1,204 @@
+
+There are several problems with the current logging mechanism that prevent end
+users from making full use of logging:
+- Message can't be redirected to different destinations on demand.
+- Lacking of a method for end users to customize messages.
+- It doesn't have a mechanism to communicate between host application and
+oe_log_message() that end users can't pass more information to the message
+function.
+
+The callback API is supposed to solve such problems.
+
+# User Experience
+
+By adding this public API, a user can customize the output of the log message:
+- The output destination can be stdout, stderr or any specified files that the
+destination is adjustable during runtime.
+- The message can be adjusted according to the log level.
+- The user can get at least as much information about the message as internel
+API. For example, log time, enclave side or host side, log level.
+- Finally, the user can define a function to customize the log message and can
+designate the funtion to process all the log messages.
+
+# Specification
+
+Several necessary items should be added to a public header file.
+- Since log level is exposed to end users, the related type should also be exposed:
+```
+typedef enum _oe_log_level
+{
+    OE_LOG_LEVEL_NONE = 0,
+    OE_LOG_LEVEL_FATAL,
+    OE_LOG_LEVEL_ERROR,
+    OE_LOG_LEVEL_WARNING,
+    OE_LOG_LEVEL_INFO,
+    OE_LOG_LEVEL_VERBOSE,
+    OE_LOG_LEVEL_MAX
+} oe_log_level_t;
+```
+- The message function declaration:
+```
+typedef void (*oe_log_callback_t)(
+    void* context,
+    bool is_enclave,
+    const char* time,
+    long int usecs,
+    oe_log_level_t level,
+    uint64_t host_thread_id,
+    const char* message);
+```
+- The necessary global context. Context may be information to be added to message, or
+just a parameter used to process the message.
+- The function to set up the callback:
+```
+oe_result_t oe_log_set_callback(void* context, oe_log_callback_t callback)
+```
+
+By including the public header file, the end user can perform customized logging.
+
+# Example
+
+Since this api just provides the log callback feature to host application, here
+the sample just shows the necessary lines of code at the host side:
+```
+#include <openenclave/host.h>
+#include <openenclave/trace.h>
+#include <stdio.h>
+
+// Include the untrusted log_callback header that is generated
+// during the build. This file is generated by calling the
+// sdk tool oeedger8r against the log_callback.edl file.
+#include "log_callback_u.h"
+
+bool check_simulate_opt(int* argc, const char* argv[])
+{
+    for (int i = 0; i < *argc; i++)
+    {
+        if (strcmp(argv[i], "--simulate") == 0)
+        {
+            fprintf(stdout, "Running in simulation mode\n");
+            memmove(&argv[i], &argv[i + 1], (*argc - i) * sizeof(char*));
+            (*argc)--;
+            return true;
+        }
+    }
+    return false;
+}
+
+void customized_log(
```

Comment:
Anand asks: 
1) Thread safety: are calls to this callback serialized?  Or can there be multiple threads calling this in parallel?
    Answer: oe_log_message already serializes messages with a mutex lock.
2) when is it safe to call set_callback?

Dave adds to point #2: can set_callback be called after enclave creation?  Can you call it multiple times and change the registered callback during processing?
    Answer: yes.   oe_log_message makes an OCALL and uses whatever callback (if any) is registered at that time

Simon asks: what is the perf overhead of oe_log_message's mutex?
Anand explains mutex is only on host side